# Segmenting and Clustering Neighborhoods in Toronto

### This notebook explores and clusters the neighborhoods in Toronto

In [1]:
# import libraries

import pandas as pd
import numpy as np

## 1. Scrape Table from Wikipedia and Pre-process

In [2]:
# scrape postal code, borough, neighborhood table for Toronto from Wikipedia

df_scrape = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

In [3]:
# multiple tables are scraped from Wikipedia page
# select the only table of interest

df_scrape = df_scrape[0]

In [4]:
df_scrape

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                         Not assigned  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
10                                        Not assigned  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                        Not assigned  
16                                        Not assigned  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                        Not assigned  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                        Not assigned  
25                                        Not assigned  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                        Not assigned  

In [5]:
df_scrape['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East Toronto         5
East York            5
Mississauga          1
Name: Borough, dtype: int64

### There are 77 Boroughs equal to 'Not assigned' to remove from the dataframe

In [6]:
# make a dataframe that does not include rows where the Borough is Not assigned

column_names = ['Postal Code', 'Borough', 'Neighborhood']

# instantiate the dataframe
df_removed_na = pd.DataFrame(columns=column_names)

# for loop to fill dataframe with any row where Borough does not equal Not assigned

for i in range(0,len(df_scrape)):
    if df_scrape.iloc[i,1] != 'Not assigned':
        df_removed_na = df_removed_na.append({'Postal Code': df_scrape.iloc[i,0], 
                                      'Borough': df_scrape.iloc[i,1], 
                                      'Neighborhood': df_scrape.iloc[i,2]}, ignore_index=True)
    elif df_scrape.iloc[i,1] == 'Not assigned':
        pass
    else:
        print('Some borough is neither equal to or not equal to "Not assigned"')                

In [7]:
print('Number of "Not assigned" rows removed:', (len(df_scrape) - len(df_removed_na)))

Number of "Not assigned" rows removed: 77


In [8]:
df_scrape.head(10)

Postal Code           Borough                                 Neighborhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

In [9]:
df_removed_na.head(6)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village

In [10]:
# check to see if any Neighborhoods are equal to 'Not assigned'

count_borough = df_removed_na['Borough'].str.count('Not assigned').sum()

count_neigh = df_removed_na['Neighborhood'].str.count('Not assigned').sum()

print('# of Boroughs with Not assigned:', count_borough)
print('# of Neighborhoods with Not assigned:', count_neigh)

# of Boroughs with Not assigned: 0
# of Neighborhoods with Not assigned: 0


### No boroughs or neighborhoods are equal to 'Not assigned' in df_removed_na

In [11]:
# check to see if all postal codes are unique

postal_code_check = df_removed_na['Postal Code'].value_counts().max()
postal_code_check

1

In [12]:
# print the shape of the final dataframe

df_removed_na.shape

(103, 3)

### The final dataframe has 103 rows (77 removed during pre-processing)

## 2. Merge Latitude and Longitude Data to Each Postal Code

In [13]:
# import latitude and longitude dataset

# sensitive code removed for sharing

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
df_lat_lon.shape

(103, 3)

In [15]:
df_removed_na.head(10)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

In [16]:
df_lat_lon.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

In [17]:
# compare length of df_removed_na to df_lat_lon

diff_in_len = len(df_removed_na) - len(df_lat_lon)
print('Difference in length:', diff_in_len)

Difference in length: 0


In [18]:
df_toronto_final = pd.merge(df_removed_na, df_lat_lon, how='inner', on='Postal Code', validate='one_to_one')
df_toronto_final.shape

(103, 5)

In [19]:
df_toronto_final.head(10)

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

## 3. Explore and Cluster the neighborhoods in Toronto

In [20]:
df_toronto_final['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

### 3a. Create a map of Toronto with neighborhoods superimposed on top.

In [21]:
# install and import folium

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.4.1               |             py_0          26 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    ------------------------------------------------------------
                       

In [22]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=toronto_location, zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto_final['Latitude'], df_toronto_final['Longitude'], df_toronto_final['Borough'], df_toronto_final['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's simplify the above map and segment and cluster only the boroughs that contain the word Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [23]:
downtown = df_toronto_final[df_toronto_final['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
central = df_toronto_final[df_toronto_final['Borough'] == 'Central Toronto'].reset_index(drop=True)
west = df_toronto_final[df_toronto_final['Borough'] == 'West Toronto'].reset_index(drop=True)
east = df_toronto_final[df_toronto_final['Borough'] == 'East Toronto'].reset_index(drop=True)

toronto_data = pd.concat([downtown, central, west, east], ignore_index=True, sort=False)

In [24]:
toronto_data.shape

(39, 5)

In [25]:
toronto_data['Borough'].value_counts()

Downtown Toronto    19
Central Toronto      9
West Toronto         6
East Toronto         5
Name: Borough, dtype: int64

In [26]:
toronto_data.head(10)

Postal Code           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M5J  Downtown Toronto   
9         M5K  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  
0                          Regent Park, Harbourfront  43.654260 -79.360636  
1        Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                           Garden District, Ryerson  43.657162 -79.378937  
3                                     St. James Town  43.651494 -79.375418  
4                                        Berczy Park  43.644771 -79.373306  
5                                 Central Bay Street  43.657952 -79.387383  
6                                           Christie  43.669542 -79.422564  
7                           Richmond, Adelaide, King  43.650571 -79.384568  
8  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
9           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576

In [27]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map of Toronto using latitude and longitude values
map_just_toronto = folium.Map(location=toronto_location, zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_just_toronto)  
    
map_just_toronto

### 3b. Use Foursquare to Collect Information on Nearby Venues

In [28]:
# define foursquare credentials and version

# sensitive code removed for sharing

In [29]:
# import libraries

import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [30]:
# function that extracts the category of the venue

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [33]:
# call foursquare API on each neighborhood

LIMIT = 100 # limit of number of venues returned by Foursquare API

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                 latitudes=toronto_data['Latitude'],
                                 longitudes=toronto_data['Longitude']
                                )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction South
Parkdale, Roncesval

In [34]:
print(toronto_venues.shape)
toronto_venues.head()

(1624, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4         Corktown Common       43.655618       -79.356211   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                 Park

In [35]:
# check how many venues were returned for each neighborhood

toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           24   
Business reply mail Processing Centre, South Ce...                     15   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     69   
Christie                                                               17   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       11   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         18   
Kensington Market, Chinatown, Grange Park                              65   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             4   
North Toronto West, Lawrence Park                                      19   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            30   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               95   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     35   
St. James Town                                                         84   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      15   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      69   
Christie                                                                17   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [36]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 225 uniques categories.


### 3c. Analyze Each Neighborhood

In [37]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# rename venue category which is also named 'Neighborhood'
toronto_onehot.rename(columns={'Neighborhood':'NeighborhoodVenue'}, inplace=True)

# method 1 - add neighborhood column back to dataframe
#toronto_onehot = pd.concat([toronto_venues['Neighborhood'], toronto_onehot], axis=1).reindex(toronto_onehot.index)

# method 2a - add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# method 2b - move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Airport  Airport Food Court  Airport Gate  \
0  Regent Park, Harbourfront        0                   0             0   
1  Regent Park, Harbourfront        0                   0             0   
2  Regent Park, Harbourfront        0                   0             0   
3  Regent Park, Harbourfront        0                   0             0   
4  Regent Park, Harbourfront        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0         0          0              0            0  
1         0          0              0            0  
2         0          0              0            0  
3         0          0              0            0  
4         0          0              0            0  

[5 rows x 226 columns]

In [38]:
toronto_onehot.shape

(1624, 226)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2   Business reply mail Processing Centre, South C...  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.058824   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12    Forest Hill North & West, Forest Hill Road Park  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                  North Toronto West, Lawrence Park  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
34                The Annex, North Midtown, Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.

In [40]:
toronto_grouped.shape

(39, 226)

Let's print each neighborhood along with the top 10 most common venues

In [41]:
num_top_venues = 10

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2            Pharmacy  0.04
3                Café  0.04
4         Cheese Shop  0.04
5  Seafood Restaurant  0.04
6            Beer Bar  0.04
7          Restaurant  0.04
8      Farmers Market  0.04
9              Bakery  0.04


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.12
1          Breakfast Spot  0.08
2   Performing Arts Venue  0.08
3             Coffee Shop  0.08
4           Grocery Store  0.04
5  Furniture / Home Store  0.04
6           Burrito Place  0.04
7            Climbing Gym  0.04
8       Convenience Store  0.04
9                     Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0                   Spa  0.07
1         Auto Workshop  0.07
2            Comic Shop  0.07
3           Pizza Place  0.07
4            Restaur

In [42]:
# function to sort the venues in descending order

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Cocktail Bar    Seafood Restaurant              Beer Bar   
1  Performing Arts Venue        Breakfast Spot           Coffee Shop   
2                 Garden            Skate Park  Fast Food Restaurant   
3         Airport Lounge      Airport Terminal               Airport   
4         Sandwich Place    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                Bakery        Farmers Market   
1           Yoga Studio                   Gym             Pet Store   
2         Auto Workshop        Farmers Market         Burrito Place   
3      Sculpture Garden           Coffee Shop         Boat or Ferry   
4                  Café       Bubble Tea Shop           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café              Pharmacy            Cheese Shop  
1             Nightclub    Italian Restaurant           Intersection  
2                   Spa           Pizza Place          Garden Center  
3                   Bar       Harbor / Marina               Boutique  
4       Thai Restaurant          Burger Joint                    Bar

### 3d. Cluster Neighborhoods

In [44]:
# import k-means from clustering stage

from sklearn.cluster import KMeans

Run *k*-means to cluster the neighborhood into 8 clusters.

In [45]:
# set number of clusters
kclusters = 8

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
#kmeans.labels_[0:10]
kmeans.labels_

array([5, 5, 0, 0, 5, 5, 5, 5, 5, 7, 0, 5, 4, 5, 5, 0, 5, 0, 6, 0, 2, 5,
       5, 5, 5, 5, 2, 1, 5, 5, 5, 5, 5, 5, 5, 3, 0, 5, 5], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [46]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               5           Coffee Shop   
1  43.662301 -79.389494               5           Coffee Shop   
2  43.657162 -79.378937               5        Clothing Store   
3  43.651494 -79.375418               5           Coffee Shop   
4  43.644771 -79.373306               5           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery        Breakfast Spot   
1                 Diner                  Park                   Bar   
2           Coffee Shop        Cosmetics Shop   Japanese Restaurant   
3                  Café            Restaurant          Cocktail Bar   
4          Cocktail Bar    Seafood Restaurant              Beer Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café                   Pub               Theater   
1              Beer Bar         Smoothie Shop        Sandwich Place   
2                 Hotel       Bubble Tea Shop                  Café   
3        Cosmetics Shop   American Restaurant              Beer Bar   
4            Restaurant                Bakery        Farmers Market   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0    Mexican Restaurant            Shoe Store                 Restaurant  
1         Burrito Place                  Café         College Auditorium  
2    Italian Restaurant                 Diner  Middle Eastern Restaurant  
3              Creperie        Lingerie Store           Department Store  
4                  Café              Pharmacy                Cheese Shop

Finally, let's visualize the resulting clusters.

In [47]:
# Matplotlib and associated plotting modules

import matplotlib.cm as cm
import matplotlib.colors as colors

In [48]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map
map_clusters = folium.Map(location=toronto_location, zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 3e. Examine Clusters

#### Cluster 1

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               0                  Café   
13  Downtown Toronto               0       Airport Service   
28      West Toronto               0                Bakery   
29      West Toronto               0                   Bar   
31      West Toronto               0       Thai Restaurant   
35      East Toronto               0      Greek Restaurant   
38      East Toronto               0    Light Rail Station   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
12           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
13        Airport Lounge      Airport Terminal                Airport   
28              Pharmacy         Grocery Store            Supermarket   
29           Men's Store            Restaurant  Vietnamese Restaurant   
31    Mexican Restaurant                  Café          Grocery Store   
35    Italian Restaurant           Coffee Shop             Restaurant   
38                Garden            Skate Park   Fast Food Restaurant   

            5th Most Common Venue   6th Most Common Venue  \
12  Vegetarian / Vegan Restaurant                  Bakery   
13               Sculpture Garden             Coffee Shop   
28                           Bank                     Bar   
29                           Café        Asian Restaurant   
31             Italian Restaurant                  Bakery   
35                 Ice Cream Shop  Furniture / Home Store   
38                  Auto Workshop          Farmers Market   

        7th Most Common Venue   8th Most Common Venue 9th Most Common Venue  \
12              Grocery Store             Pizza Place                   Bar   
13              Boat or Ferry                     Bar       Harbor / Marina   
28                       Café  Furniture / Home Store           Music Venue   
29                Yoga Studio                 Brewery           Record Shop   
31  Cajun / Creole Restaurant               Speakeasy   Fried Chicken Joint   
35    Fruit & Vegetable Store                     Pub           Pizza Place   
38              Burrito Place                     Spa           Pizza Place   

   10th Most Common Venue  
12                   Park  
13               Boutique  
28                Brewery  
29            Pizza Place  
31   Fast Food Restaurant  
35                 Lounge  
38          Garden Center

#### Cluster 2

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20           Yoga Studio          Dance Studio     Electronics Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20               Dog Run   Distribution Center         Discount Store

#### Cluster 3

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               2                  Park   
26   Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Trail          Cupcake Shop   
26            Playground                 Trail          Tennis Court   

          5th Most Common Venue        6th Most Common Venue  \
14  Eastern European Restaurant                   Donut Shop   
26                 Cupcake Shop  Eastern European Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14      Doner Restaurant               Dog Run   Distribution Center   
26            Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
14         Discount Store  
26    Distribution Center

#### Cluster 4

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
34  East Toronto               3                 Trail     NeighborhoodVenue   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
34     Health Food Store                   Pub        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
34         Deli / Bodega      Department Store          Dessert Shop   

   9th Most Common Venue 10th Most Common Venue  
34                 Diner            Yoga Studio

#### Cluster 5

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22         Jewelry Store                 Trail      Sushi Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
22          Dance Studio  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 6

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               5           Coffee Shop   
1   Downtown Toronto               5           Coffee Shop   
2   Downtown Toronto               5        Clothing Store   
3   Downtown Toronto               5           Coffee Shop   
4   Downtown Toronto               5           Coffee Shop   
5   Downtown Toronto               5           Coffee Shop   
6   Downtown Toronto               5         Grocery Store   
7   Downtown Toronto               5           Coffee Shop   
8   Downtown Toronto               5           Coffee Shop   
9   Downtown Toronto               5           Coffee Shop   
10  Downtown Toronto               5           Coffee Shop   
11  Downtown Toronto               5                  Café   
15  Downtown Toronto               5           Coffee Shop   
16  Downtown Toronto               5                Bakery   
17  Downtown Toronto               5           Coffee Shop   
18  Downtown Toronto               5           Coffee Shop   
23   Central Toronto               5        Clothing Store   
24   Central Toronto               5                  Café   
25   Central Toronto               5           Pizza Place   
27   Central Toronto               5           Coffee Shop   
30      West Toronto               5                  Café   
32      West Toronto               5             Gift Shop   
33      West Toronto               5                  Café   
36      East Toronto               5                  Park   
37      East Toronto               5                  Café   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Park                Bakery        Breakfast Spot   
1                   Diner                  Park                   Bar   
2             Coffee Shop        Cosmetics Shop   Japanese Restaurant   
3                    Café            Restaurant          Cocktail Bar   
4            Cocktail Bar    Seafood Restaurant              Beer Bar   
5          Sandwich Place    Italian Restaurant   Japanese Restaurant   
6                    Café                  Park           Candy Store   
7                    Café            Restaurant                 Hotel   
8                Aquarium                  Café                 Hotel   
9                   Hotel                  Café           Salad Place   
10             Restaurant                  Café                 Hotel   
11                 Bakery                   Bar   Japanese Restaurant   
15                   Café                 Hotel    Italian Restaurant   
16            Coffee Shop            Restaurant                  Café   
17                   Café                 Hotel                   Gym   
18    Japanese Restaurant      Sushi Restaurant            Restaurant   
23            Coffee Shop           Yoga Studio    Chinese Restaurant   
24         Sandwich Place           Coffee Shop                  Park   
25         Sandwich Place          Dessert Shop                  Café   
27                    Pub    Light Rail Station                  Bank   
30  Performing Arts Venue        Breakfast Spot           Coffee Shop   
32         Breakfast Spot               Dog Run    Italian Restaurant   
33       Sushi Restaurant           Coffee Shop           Pizza Place   
36                    Pub        Sandwich Place         Burrito Place   
37            Coffee Shop             Gastropub               Brewery   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                   Café                        Pub               Theater   
1               Beer Bar              Smoothie Shop        Sandwich Place   
2                  Hotel            Bubble Tea Shop                  Café   
3         Cosmetics Shop        American Restaurant              Beer Bar   
4             Restaurant                     Bakery        Farmers Market   
5                   Café            Bub

#### Cluster 7

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               6                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Swim School              Bus Line         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
19     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 8

In [56]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 7, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21  Central Toronto               7                 Hotel   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21        Breakfast Spot                  Park                   Gym   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21  Gym / Fitness Center           Pizza Place      Department Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21     Food & Drink Shop          Dance Studio         Sandwich Place

### 3f. Cluster Observations

These are the trend of each cluster:

Cluster 1 - Nearby region that includes west and east of very downtown Toronto. Contains transportation (airport, lightrail) and secondly resturants and bars.

Cluster 2 - Far north/central of downtown. Small cluster with garden, yoga and dance studio.

Cluster 3 - Immediately north/central of downtown. Small cluster based on parks, trails and playgrounds.

Cluster 4 - East Toronto beaches neighborhood. Small cluster with trails, health food stores and department stores.

Cluster 5 - North/central of downtown. Small cluster with parks, trails and resturants.

Cluster 6 - Downtown Toronto cluster. Largest cluster with predominantly coffee shops/cafe, resturants and bars.

Cluster 7 - Farthest north/central of downtown. Small cluster with parks, swim schools and transportation (bus line).

Cluster 8 - Far north/central of downtown. Small cluster with hotels, parks and resturants.